lib/pmap.dart

```diff
@@ -1,109 +1,178 @@
+    required this.receivePort,
+  });
+
+  static Future<_ProcessorIsolate<T, U>> spawn<T, U>(
+      U Function(T input) mapper) async {
+    final receivePort = ReceivePort();
+    final isolate = Isolate.spawn(
+      _process,
+      _Processor(mapper, receivePort.sendPort),
+    );
+    return _ProcessorIsolate(isolate: await isolate, receivePort: receivePort);
   }
 }
 
-/// Operates like `Iterable.map` except performs the function `mapper` on a
-/// background isolate.  `parallel` denotes how many background isolates to use.
+/// Operates like [Iterable.map] except performs the function [mapper] on a
+/// background isolate. [parallel] denotes how many background isolates to use.
 ///
-/// This is only useful if the computation time of `mapper` out paces the
+/// This is only useful if the computation time of [mapper] out paces the
 /// overhead in coordination.
 ///
-/// Note: `mapper` must be a static method or a top-level function.
-Stream<U> pmap<T, U>(Iterable<T> list, U Function(T input) mapper,
-    {int parallel = 1}) {
-  List<_ProcessorIsolate<T, U>> isolates = [];
-  StreamController<U> controller = StreamController<U>();
-  Iterator<T> it = list.iterator;
-  int sendCount = 0;
-  int receiveCount = 0;
-  List<_Enumerated<U>> buffer = [];
-  for (int i = 0; i < parallel; ++i) {
-    _ProcessorIsolate<T, U> isolate = _ProcessorIsolate<T, U>();
-    isolate.processor.mapper = mapper;
-    isolates.add(isolate);
-    isolate.spawn().then((x) async {
-      isolate.receivePort.listen((dynamic result) {
-        if (isolate.sendPort == null) {
-          isolate.sendPort = result;
-          isolate.sendPort.send(isolate.processor);
-          if (it.moveNext()) {
-            isolate.sendPort.send(_Enumerated(index:sendCount++, value:it.current));
-          } else {
-            isolate.completer.complete();
-          }
-        } else {
-          _Enumerated<U> enumeratedResult = result;
-          if (enumeratedResult.index == receiveCount) {
-            controller.add(enumeratedResult.value);
-            receiveCount++;
+/// If preserving the order of the [iterable] is not required,
+/// the [preserveOrder] setting can be set to `false`, reducing memory and
+/// returning results as soon as available.
+///
+/// Note: [mapper] must be a static method or a top-level function.
+Stream<U> pmap<T, U>(
+  Iterable<T> iterable,
+  U Function(T input) mapper, {
+  int parallel = 1,
+  bool preserveOrder = true,
+}) {
+  assert(
+    parallel > 0,
+    'There need to be at least one worker, but got $parallel',
+  );
+
+  final controller = StreamController<U>();
+
+  // If the iterable was a list, the `currentIterableIndex` would be the
+  // index of the element in `iterable.iterator`.
+  var currentIterableIndex = -1;
+  final it = iterable.iterator;
+  bool nextInputElement() {
+    final hasNext = it.moveNext();
+    if (hasNext) currentIterableIndex++;
+    return hasNext;
+  }
+
+// The number of already emitted elements
```

Comment:
nit: leading spaces here and a couple lines down.

Reply:
Thanks, I trusted in `dartfmt`, but for comments it was not strict enough.